src/pyk/konvert.py

```diff
@@ -1,61 +1,166 @@
-from collections import defaultdict
-from dataclasses import dataclass
-from functools import cached_property
-from pathlib import Path
-from typing import Dict, Final, FrozenSet, Optional, Set, Union, final
-
-from .cli_utils import check_dir_path, check_file_path
-from .kore.parser import KoreParser
-from .kore.syntax import Definition, Sort
+from functools import reduce
+from typing import Final, Optional, Tuple
+
+from .kast.inner import KApply, KInner, KSequence, KSort, KToken, KVariable
+from .kast.outer import KDefinition
+from .kore.kompiled import KompiledKore
+from .kore.syntax import DV, App, EVar, MLPattern, MLQuant, Pattern, Sort, SortApp, String
+from .prelude.bytes import BYTES
+from .prelude.k import K
+from .prelude.string import STRING
 from .utils import FrozenDict
 
+# ------------
+# KAST-to-KORE
+# ------------
+
+ML_CONN_LABELS: Final = {
+    '#Top': r'\top',
+    '#Bottom': r'\bottom',
+    '#Not': r'\not',
+    '#And': r'\and',
+    '#Or': r'\or',
+    '#Implies': r'\implies',
+    '#Iff': r'\iff',
+}
+
+ML_QUANT_LABELS: Final = {
+    '#Exists': r'\exists',
+    '#Forall': r'\forall',
+}
+
+ML_PRED_LABELS: Final = {
+    '#Ceil': r'\ceil',
+    '#Floor': r'\floor',
+    '#Equals': r'\equals',
+    '#In': r'\in',
+}
+
+ML_PATTERN_LABELS: Final = dict(
+    **ML_CONN_LABELS,
+    **ML_QUANT_LABELS,
+    **ML_PRED_LABELS,
+)
+
+
+def kast_to_kore(
+    kast_defn: KDefinition,
+    kompiled_kore: KompiledKore,
+    kast: KInner,
+    sort: Optional[KSort] = None,
+) -> Pattern:
+    if sort is None:
+        sort = K
+    kast = kast_defn.sort_vars(kast, sort)
+    kore = _kast_to_kore(kast)
+    return kompiled_kore.add_injections(kore, _ksort_to_kore(sort))
+
+
+def _kast_to_kore(kast: KInner) -> Pattern:
+    if type(kast) is KToken:
+        return _ktoken_to_kore(kast)
+    elif type(kast) is KVariable:
+        return _kvariable_to_kore(kast)
+    elif type(kast) is KSequence:
+        return _ksequence_to_kore(kast)
+    elif type(kast) is KApply:
+        return _kapply_to_kore(kast)
+
+    raise ValueError(f'Unsupported KInner: {kast}')
+
+
+def _ktoken_to_kore(ktoken: KToken) -> DV:
+    token = ktoken.token
+    sort = ktoken.sort
+
+    if sort == STRING:
+        assert token.startswith('"')
+        assert token.endswith('"')
+        return DV(_ksort_to_kore(sort), String(token[1:-1]))
+
+    if sort == BYTES:
+        assert token.startswith('b"')
+        assert token.endswith('"')
+        return DV(_ksort_to_kore(sort), String(token[2:-1]))
+
+    return DV(_ksort_to_kore(sort), String(token))
+
+
+def _ksort_to_kore(ksort: KSort) -> SortApp:
+    return SortApp('Sort' + ksort.name)
+
+
+def _kvariable_to_kore(kvar: KVariable) -> EVar:
+    sort: Sort
+    if kvar.sort:
+        sort = _ksort_to_kore(kvar.sort)
+    else:
+        sort = SortApp('SortK')
+    return EVar('Var' + munge(kvar.name), sort)
+
+
+def _ksequence_to_kore(kseq: KSequence) -> Pattern:
+    if not kseq:
+        return App('dotk')
```

Comment:
```suggestion
    if len(kseq) == 0:
        return App('dotk')
```

I prefer this more explicit check.

Reply:
Same elsewhere, I prefer not relying on "truthiness" of values, simplify because then I can't tell looking at this line if it's something that I should expect to be an optional type or a iterable/list-ish type.

Reply:
This is a recommendation in [PEP8](https://peps.python.org/pep-0008/#programming-recommendations).

```python3
# Correct:
if not seq:
if seq:
```

```python3
# Wrong:
if len(seq):
if not len(seq):
```

But I see your point: if `kseq` is `Optional`, then `mypy` flags `len(kseq) == 0` as a type error, so the reviewer can be confident that there is no bug coming from an unhandled `None` case. In this case though, I think the context is simple enough to follow the recommendation.